### ISSUE TITLE
[FEATURE] Standardising the Dexios Header Sizes

### ISSUE THREAD

**Issue by brxken128** [brxken128/dexios#15]
In the next major release of Dexios, I think headers should be standardised to 48 (or 64) bytes at the start of the file. This would be able to account for the 16 byte salt, and the (current) max nonce length of 24 bytes, with room to spare.

The header would then be padded with random bytes/zeroes until it reaches that 48 byte length. This would mean there is no risk of data loss when stripping/dumping the headers, as the start of the encrypted data would always be after the 48 byte header.

We could possibly use these extra bytes to store information about the cipher and mode used (e.g. XChaCha20-Poly1305/AES-256-GCM and memory/stream modes). This would remove the need for the user to even care about what mode their file was encrypted with, and Dexios would be able to detect it automatically.

This would break compatibility with older files, so the major version would **have** to be incremented, in order to adhere to semantic versioning.

**Comment by brxken128** [comment]
@Rust-Galt do you think this is worth it?

I feel as though it is, but I'm hesitant on bumping up to v8.0.0 just for this feature.

It would make everything a lot more user-friendly as they wouldn't have to specify `-m` or `-s` in header functions, or for decrypting.

**Comment by brxken128** [comment]
I think a header size of 64 bytes would work best.

2 bytes can be allocated towards denoting the encryption algorithm (such as XChaCha20-Poly1305)
2 bytes can be allocated towards denoting the mode that it was encrypted in (stream/memory)

16 bytes of that can be salt, with room to expand.

Our current max nonce size is for XChaCha20-Poly1305 in memory mode, at 24 bytes

This brings us up to 44 bytes, with 20 to spare. I feel as though this is a good trade-off between header size and room to expand - we don't need to allocate ourselves too much data.

I think we should have one standard function for reading headers. It should take in an `&mut File`

* It will read the first two bytes, and get the encryption algorithm
* It will read the next two bytes, and get the encryption mode
* It will read the next 16 bytes, and get the salt
* It will read the next 16 bytes, and discard them as they should be zeroes
* It will finally read the last X bytes, depending on encryption mode/algorithm
* It should read enough extra bytes to make the file handle seek to 64 bytes, no matter the nonce length

I'm not too sure about the extra empty space - maybe we could start using a 32-byte salt? I don't see what benefits that would provide us though.

I don't think there's any AEAD that we could implement which requires larger than a 24 byte nonce, so that shouldn't be an issue.

**Comment by brxken128** [comment]
> I think we should have one standard function for reading headers. It should take in an `&mut File`

This would allow for us to pass a detached header file directly, too.

We would also need a standard function that takes an `&mut File` and writes a header to it - it would also need to take information such as encryption mode/algorithm, and stuff like that.

Maybe even another function could handle serialisation/deserialisation of those types, so they're all in one place and can easily be expanded upon. I quite like the idea of hex.

I think there should be a switch for encrypting a file with a detached header, such as `--detached file.header`.

This will write the header to the specified file. It will then write 64 empty bytes + the encrypted data to the output file. The user would be required to provide that detached header on decryption, as there would be no header information for Dexios to read otherwise.

If we keep 64 bytes spare at the start of the file, empty or not, the user has the room to restore the header at will.

Dexios encrypted information should always start after the first 64 bytes of the file, imo. It keeps everything standardised.

**Comment by brxken128** [comment]
I've started working on a decent implementation - I have left out argon2 parameters though, I honestly don't know enough about them.

I do like the idea of a `header/app_version` prefix - I've placed all of the header functions in one place so this should be fairly easy to add.

> I quite like the idea of hex.

I was referring to using hex bytes to identify, e.g. `[0x0E, 0x01]` - `0x0E` being encryption algorithm and `0x01` being XChaCha20-Poly1305.

My plan wasn't to implement support for encrypting *with* a detached header, but instead detaching it from the file during encryption (padding the start of the file with 64 empty bytes instead). For users who plan to strip their header and store it elsewhere, it could've been useful, but `strip` just provides that functionality anyway.

I'll implement versioning within the header once I've finished this - I'm moving both `encrypt` modes over to use the new header functions, and I'm rewriting memory mode because I'm not happy with how I had the function structured.
